roachtest: eliminate the concept of subtests

andreimatei opened the pull request:
The point of subtests was that all the subtests under a root share a
cluster. This patch introduces more flexible policies for sharing
clusters across tests (only the declaration of these policies is here;
the implementation is in future commits). The policies replace the need
for subtests, so they go away, simplifying the code considerably.

Release note: None

andreimatei committed:
roachtest: eliminate the concept of subtests

The point of subtests was that all the subtests under a root share a
cluster. This patch introduces more flexible policies for sharing
clusters across tests (only the declaration of these policies is here;
the implementation is in future commits). The policies replace the need
for subtests, so they go away, simplifying the code considerably.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/34071)
<!-- Reviewable:end -->

andreimatei commented:
Peter, this is part of the old PR for rewriting the roachtest test runner.
I'm not sure I can merge this in isolation and also I've rebased it hastily. For now what I'm interested in is relitigating what the test's spec should look like wrt cluster reuse.
  Pedantically, I think the most flexible way to model this would be to have a way for tests to say what cluster they'll accept and separately a different way for them to say what state they intend to leave the cluster in. This is not what the patch implements, though. I implement something simpler that models a virus' interactions with hosts: a virus can infect clean hosts, it is a no-op for hosts that are already carriers, and it can kill a host. Each test declares what virus it carries, if any(*). I think it's good and EZ.

When this came up before, you wanted to "fold this (spec) into the nodeSpec". If you still feel that way, please fill in some details and explain it to me again. What exactly would we put and where? testSpec.Nodes is an array. How would we express "nobody can reuse this cluster after me"? You recognized that the scheme I've implemented has the advantage that a cluster can go from clean->tagged, but you weren't impressed.
  

(*) To make the analogy work, we say that our deadly viruses kill a host immediately, so there's no point in differentiating between different deadly viruses.

petermattis reviewed (commented):
My "fold this into the nodeSpec" idea is as follows. Rather than adding a `ClusterReusePolicy` field, we add a `reuseNeverOption` and `reuseTaggedOption`:

```
type reuseNeverOption struct{}

var reuseNever reuseNeverOption

func (reuseNeverOption) apply(spec *nodeSpec) {
  spec.reusePolicy = reuseNever
}

type reuseTaggedOption string

func reuseTagged(s string) reuseTaggedOption {
  return reuseTaggedOption(s)
}

func (o reuseTaggedOption) apply(spec *nodeSpec) {
  spec.reusePolicy = o
}
```

We'd have to add a `nodeSpec.reusePolicy` field, presumably with some interface type that the options above implement. I'll leave those details to you.

The default reuse policy would be "any". In a `testSpec`, this would be specified like:

```
testSpec{
	Name:       fmt.Sprintf("kv%d/encrypt=%t/nodes=%d", p, e, n),
	Nodes:      nodes(n+1, cpu(8), reuseTagged("nobarrier")),
}
```

This suggestion is purely about ergonomics. I think the functionality would be the same. When a cluster is created, you'd copy the reuse policy from the test into the cluster. When looking for an existing cluster, you'd match on the reuse policy. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34071#-:-LWRCsWr8ntK4BcskXzx:bbtghvs)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cmd/roachtest/jepsen.go, line 296 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34071#-LWRBaTH2qkeDLu1U1bl:-LWRBaTH2qkeDLu1U1bm:bb7tcuu) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f5432714d9fb153dfee61eae1312574220cd730/pkg/cmd/roachtest/jepsen.go#L296)):*
> ```Go
> 				// if they detect that the machines have already been properly
> 				// initialized.
> 				ClusterReusePolicy: OnlyTagged("jepsen"),
> ```

So the jepsen tests can reuse an existing 6-node cluster without a tag, or a 6-node cluster that is tagged `jepsen`? One caveat I see with this approach is that we might end up regenerating the `jepsen` clusters repeatedly depending on how the jepsen tests get scheduled with other tests. Prior to this PR, we would run all of the jepsen tests in a group sequentially, so we'd only end up creating (and initializing) 3 `jepsen` clusters. Do you have sense of what will happen with this PR? 


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
well, but `nodes(...)` returns `[]NodeSpec`, so... what do the `apply()` methods get called with? With what I have in mind / with the current roachprod infra, it wouldn't be possible to reuse some nodes in a cluster but not others, so what do you do with inconsistent reuse policies across `NodeSpecs`? I know that currently the `nodes()` method only ever returns an array with a single `NodeSpec`, and I know (or rather I guess) that no test combines the output of multiple `nodes()` call, but I think we should leave that window open (I actually created a branch and started changing things to have everybody use `NodeSpec` instead of `[]NodeSpec`, so I tried to be the bigger man, but I went back). So what this says is that the reuse policy and the `CPUs` field, for example, don't belong in the same struct. The `reusePolicy` belongs in something that refers to a cluster, not a node.

So, you say you'll leave the details to me, but if you leave them to me, you end up with this patch :)
Another thing about your proposal is that it all suggests that you want the reuse policy to have a good default - (any?), but I think that'd be bad. I'd rather force tests to declare it explicitly, cause it's dangerous.
So... maybe you find this acceptable after all?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34071#-:-LWSYb8B0f2H08asdlcD:b18cmr2)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cmd/roachtest/jepsen.go, line 296 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34071#-LWRBaTH2qkeDLu1U1bl:-LWSX-e34v4Dzl1HJppA:b-xsz2n0) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f5432714d9fb153dfee61eae1312574220cd730/pkg/cmd/roachtest/jepsen.go#L296)):*
> So the jepsen tests can reuse an existing 6-node cluster without a tag, or a 6-node cluster that is tagged jepsen?

correct

> ...

yeah, so that's why I probably can't merge this alone.
In the test runner rewrite, the scheduling policy is that you start picking up work and allocating resources for that work randomly, until you run out of resources for starting new clusters. Then, as tests finish, they give you back clusters and then for each cluster you chose the best work that can use it. For an untagged cluster, that's other work that leaves it untagged or, if no such thing, then a test with a random tag and finally a non-reuse test. For a tagged cluster, it's another test tagged the same or bust.
So in practice this will mean that the number of Jepsen clusters we'll create is gonna depend on how long different tests take. If the Jepsen ones are relatively fast, we'll probably create a single one. If they're really slow, we might create many (which is good).


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
`nodes()` currently always returns a `[]nodeSpec` with exactly 1 element. I'd be fine with removing that generality and having it return a `*nodeSpec`. And perhaps `nodeSpec` should be renamed to `clusterSpec`. The thinking around having a `[]nodeSpec` was never finished, and I doubt we're going to do anything on it soon.

> So, you say you'll leave the details to me, but if you leave them to me, you end up with this patch :)
Another thing about your proposal is that it all suggests that you want the reuse policy to have a good default - (any?), but I think that'd be bad. I'd rather force tests to declare it explicitly, cause it's dangerous.
So... maybe you find this acceptable after all?

Can you elaborate on why `any` reuse is a bad default? Dangerous because you might accidentally reuse a cluster? I think the majority of tests don't care about cluster reuse. I don't expect that to change. Do you?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34071#-:-LWSk7Ad1A8WNGbI0NTf:bt44b16)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
ok, did your thing. I just don't have a good model to predict what fights you're going to take :P

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34071#-:-LWTN60b7OIql2udoTBf:b-cwbvdd)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> ok, did your thing. I just don't have a good model to predict what fights you're going to take :P

Unpredictable by nature.

This is looking good to me. I started to review the change, but then saw that you need to finish rebasing and do various cleanups. Ping me again when this is ready for another look.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34071#-:-LWWKVM48xEmqbrIA7gL:bvw43ru)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @petermattis)

___
*[pkg/cmd/roachtest/test.go, line 104 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34071#-LWWL1Bg7S9C2aSClHdy:-LWWL1Bg7S9C2aSClHdz:b-s6hwa6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b037dd9e3d097cc540878d8d7e472ac5b0ec3319/pkg/cmd/roachtest/test.go#L104)):*
> ```Go
> 	onlyTagged
> 	any
> )
> ```

I think you should move these defs into `cluster.go`.


<!-- Sent from Reviewable.io -->
